#166789989 Users should receive a JWT on Successful Registration.

elemanhillary-zz opened the pull request:
#### What does this PR do?
- Adds jsonwebtoken strategy for authentication
- Adds registration and login endpoint

#### Description of Task to be completed
- Enable user to register and login
- Enable user authentication with jwt(jsonwebtoken) authentication
#### How should this be manually tested?
1. Run the following commands in Terminal
```
git clone https://github.com/andela/ah-92explorers-backend.git
cd ah-92explorers-backend
sudo mv .env.example .env
npm install
sequelize db:migrate
npm start
```
2. Open your favourite browser and navigate to the following:
```
localhost:3000/api/users
localhost:3000/api/users/login
```

#### Any background context you want to provide?
- None

#### What are the relevant pivotal tracker stories?
[166789989](https://www.pivotaltracker.com/story/show/166789989)

#### Screenshots (if appropriate)
![Imgur](https://i.imgur.com/wdd0nEp.png)

#### Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I don't have more than 3 major commits on this PR

Commit:
feature(create-crud-superadmin): create and implement crud for superadmin

- create class to implement crud operations
- store user data in database
- setup and carryout all unit tests for the operation

[Finishes #166790016]

kodek-sleuth reviewed (changes requested):
Please don't return the success key when returning an error even if its 0
Please use message key not msg

kodek-sleuth commented on tests/auth.test.js:
```diff
@@ -0,0 +1,105 @@
+const email = faker.internet.email();
+const userName = faker.internet.userName();
+
+const dummyData = {
```
we are going to use seeders for storing dummy data

kodek-sleuth reviewed (changes requested):
Please don't return a password/id in the response and I would prefer if we used UUID to generate id's

niyoceles reviewed (dismissed):
The test is not passing on Travis CI, please rework on it.

elemanhillary-zz commented:
> The test is not passing on Travis CI, please rework on it.

Tests are now passing @niyoceles

kodek-sleuth commented on .travis.yml:
```diff
@@ -1,7 +1,11 @@
+  - npm i sequelize -g
+  - sequelize db:migrate
 after_success:
-  - npm test
```
This command should be run in the script not here. It completely overlooks the presence of the latter

elemanhillary-zz replied:
@kodek-sleuth kindly learn how to review, code which is in a red line indicates it has been deleted and changed with a new piece of code.

kodek-sleuth commented on migrations/20190704183748-create-user.js:
```diff
@@ -0,0 +1,39 @@
+module.exports = {
```
Use es6 here, @slack here

kodek-sleuth reviewed (changes requested):
Please follow all eslint rules

IsaiahRn reviewed (dismissed):
You have implemented authentication and jwt strategy for authentication

elemanhillary-zz commented on .editorconfig:
```diff
@@ -0,0 +1,8 @@
+root = true
```
remove `.editorconfig` from version control

elemanhillary-zz commented on .eslintignore:
```diff
@@ -0,0 +1,2 @@
+node_modules
+!.eslintrc
```
remove `!.eslintrc`

elemanhillary-zz commented on .eslintrc:
```diff
@@ -15,14 +15,14 @@
         "sourceType": "module"
```
use `0` to turn off some eslint rules

elemanhillary-zz commented on controllers/auth.js:
```diff
@@ -0,0 +1,60 @@
+  const {
+    username, firstName, lastName, email,
+  } = req.body;
+  const password = Auth.hashPassword(req.body.password);
```
1. destructure password in the request body on line `6`
2. rename `password` to `hashedPassword`

elemanhillary commented on controllers/auth.js:
```diff
@@ -0,0 +1,60 @@
+    });
+    await transaction.commit();
+    if (newUser) {
+      return res.status(201).json({
```
restructure:
```
{
message,
   user: {
       token,
       username,
       email,
     }
}
```

elemanhillary-zz commented on controllers/auth.js:
```diff
@@ -0,0 +1,60 @@
+    if (newUser) {
+      return res.status(201).json({
+        success: 1,
+        token: Auth.genToken(username, firstName, lastName, email),
```
change token to take these params
```
Auth.genToken(username, email)
```

elemanhillary-zz commented on controllers/auth.js:
```diff
@@ -0,0 +1,60 @@
+    const user = await db.User.findOne({ where: { email } }, { transaction });
+    await transaction.commit();
+    if (!user) {
+      return res.status(401).json({ message: 'user doesnot exit' });
```
fix typos `exit` to `exist`

elemanhillary-zz commented on controllers/auth.js:
```diff
@@ -0,0 +1,60 @@
+    if (passBool) {
+      return res.status(200).json({
+        success: 1,
+        token: Auth.genToken(user.username, user.firstName, user.lastName, user.email),
```
destructure user object return by the db

elemanhillary-zz commented on helpers/auth.js:
```diff
@@ -0,0 +1,27 @@
+    return bcrypt.compareSync(password, hashedPassword);
+  }
+
+  static genToken(username, firstname, lastname, email) {
```
genToken shoud take `username, email` as params

elemanhillary-zz commented on middlewares/checkToken.js:
```diff
@@ -0,0 +1,35 @@
+  } else {
+    return res.status(401).json({
+      success: false,
+      message: 'token empty',
```
change `token empty` to `missing authorization token` and json key` message` to `error`

kodek-sleuth commented on middlewares/checkToken.js:
```diff
@@ -0,0 +1,35 @@
+    });
+  } else {
+    return res.status(401).json({
+      success: false,
```
If you sending an error key, we should not use a success key

kodek-sleuth commented on routes/api/index.js:
```diff
@@ -1,8 +1,8 @@
 const router = express.Router();
 
-router.use('/', welcome);
+router.use('/', routes);
```
please use users instead of routes

IsaiahRn reviewed (dismissed):
It can be merged!

kodek-sleuth commented on .gitignore:
```diff
@@ -1,10 +1,11 @@
 
 npm-debug.log*
-
+seeders
```
Remove seeders from the gitignore file

kodek-sleuth replied:
Remove seedeers from the gitignore file

kodek-sleuth commented on .sequelizerc:
```diff
@@ -1,5 +1,4 @@
-
 const path = require('path');
 
 module.exports = {
```
Use es6 to export this

kodek-sleuth commented on controllers/auth.js:
```diff
@@ -0,0 +1,59 @@
+  const { email, password } = req.body;
+  const transaction = await db.sequelize.transaction();
+  try {
+    const user = await db.users.findOne({ where: { email: req.body.email } }, { transaction });
```
This users object can be destructured from the db object you imported e.g 

`import db from ../models`;
const { users } = db;

kodek-sleuth commented on tests/token.test.js:
```diff
@@ -0,0 +1,51 @@
+    expect(next.called).to.equal(false);
+  });
+
+  it('request should contain user info if good token was provided', () => {
```
begin your it blocks with should

elemanhillary committed:
feat(setup-jwt) ensure user receives jwt

- ensure user registers
- ensure user signin
- ensure user receives jwt on successful registration

[Maintains #166789989]

IsaiahRn reviewed (dismissed):
It is ready to be merged!

kodek-sleuth commented on controllers/auth.js:
```diff
@@ -0,0 +1,59 @@
+    const user = await db.users.findOne({ where: { email: req.body.email } }, { transaction });
+    await transaction.commit();
+    if (!user) {
+      return res.status(404).json({ message: 'user doesnot exist' });
```
I think you should re-edit this message to `user not found`

kodek-sleuth commented on controllers/auth.js:
```diff
@@ -0,0 +1,61 @@
+    username, firstname, lastname, email, password,
+  } = req.body;
+  const hashedPassword = Auth.hashPassword(password);
+  const transaction = await db.sequelize.transaction();
```
restructure this line please

kodek-sleuth reviewed (approved):
This branch is good to be merged

IsaiahRn reviewed (approved):
It can be merged!
